docs/community/20241112-node-bootstrapping.md

```diff
@@ -0,0 +1,90 @@
+---
+title: ClusterAPI Node Bootstrapping Working Group
+authors:
+    - "@t-lo"
+reviewers:
+    - "@elmiko"
+    - "@eljohnson92"
+    - "@johananl"
+    - "@tormath1"
+
+creation-date: 2024-11-12
+last-updated: 2024-11-12
+status: proposed
+see-also:
+  - https://github.com/kubernetes-sigs/cluster-api/issues/5294
+  - https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit
+  - https://github.com/kubernetes-sigs/cluster-api/issues/9157
+  - https://youtu.be/0AhA4Box3MM?si=IfKJfMWmlA9EW7ri&t=172
+---
+
+# Node Bootstrapping Working Group ("CAPI-NoBo")
+
+This document briefly outlines the scope, communication media, and
+stakeholders for a Working Group dedicated to the Provisioning aspect of
+Node Bootstrapping.
+
+Provisioning, in this context, refers to configuring and customising the
+node operating system to prepare it to serve as a ClusterAPI worker cluster node.
+
+
+## User Story and Problem Statement
+
+As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
+support of ClusterAPI in my distribution, I would like to improve ClusterAPI's
+node bootstrapping process and provide specifications on ClusterAPI's expectations
+and requirements for provisioning-time configuration and customisation.
+
+Initial focus of the working group is on separating node bootstrapping and node
+OS configuration / customisation ("node provisioning" from here on).
+At the time of writing, bootstrapping and provisioning are architecturally
+intertwined; both are implemented in the bootstrap provider.
+This increases maintenance effort and effectively prevents re-using provisioning
+features across bootstrap providers.
```

Comment:
I think that this statement is somehow more hinting a solution (separation) than a problem.

Base on my understanding, we are trying to solve two problems:
- to offer to the users more combinations of provisioner/bootstrappers.
- to ensure those combinations can evolve freely, taking befits from the greatest an the latest the underlying technologies

But IMO, users don't care about separation at all. 
In fact, users will always need both a provisioner and a bootstrapper to get a working cluster.

If instead we split provisioner and a bootstrapper into two separate components, probably we are not making a good service both to the users, nor to ourself (as project maintainers), nor to the community that builds on top of Cluster API:

- we are adding to the system another dimension of complexity, that makes this change highly disruptive for users
- we are imaging a change that will impact not only the users, but the entire of the CAPI ecosystem, think e.g. to the CAPI operator, or all the products built on top of Cluster API.
- we are adding to the users an additional operational burden (they have to take care of yet another component/provider, its upgrades, its compatibility matrix, its CVEs etc etc). 
- we are imaging a change so complex and extensive that, being very pragmatic, the chances that this change might happen drastically drop (e.g. it will impact clusterctl, clusterclass, runtime extensions, all the controllers, almost everything; in fact this is why this problem is laying around for so long)

Frankly speaking, this seems too much to digest for everyone. 

The good news, it that I think that it is possible to look at the problem above from a different angle and find a better solution.

How to solve the two problem above (offer more combinations/ensure those combinations can evolve freely) without introducing separation?

That's up to the working group to figure out, but If I can give my two cents, what we are looking for is a solution that doesn't impact users, but makes it easier to implement and evolve different provisioners (e.g. we want to bump ignition!), while re-using existing bootstrappers (e.g. kubeadm).

I think that the key point in the sentence above is software re-use

So, a possible solution might be in a combination of federated development/separated repositories (providers as we call it in CAPI) + something that makes it easy to reuse code e.g. making kubeadm provider a library (details TBD of course, but I hope you get the idea)

Please also see https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit?disco=AAABX57uOnU where I make other considerations about the same point

Happy to chat about it if you are interested on this angle

Reply:
These are good points - will rework the proposal accordingly. Admittedly I wrote the first draft with the existing proposal in mind.

Thank you for your review!

Reply:
Thanks @fabriziopandini. I completely understand the concerns you've raised and am likewise worried by them. I'm also 100% for maximizing code reuse.

If I understand correctly, what you are proposing is to keep provisioners (e.g. Ignition) tightly coupled with bootstrappers (e.g. kubeadm) and offer multiple bootstrap providers for each bootstrapper, one for each provisioner. So, for example, we would have a **kubeadm + cloud-config** bootstrap provider and a **kubeadm + Ignition** bootstrap provider in separate repositories.

If that's what you meant, I'm seeing the following problems with this approach:

1. The cartesian product of bootstrappers and provisioners can grow very large. True, with 1 bootstrapper and 2 provisioners we would have only 2 repositories to maintain, but what if we have 3 bootstrappers and 3 provisioners? That's **9 repositories** already. AFAICT this isn't premature optimization since we seem to already have [8 supported bootstrappers](https://cluster-api.sigs.k8s.io/reference/providers.html?highlight=k3s#control-plane) and at least 3 provisioners (cloud-init and Ignition, with CloudBase Init either implemented or planned - not sure). That's 24 repos if we don't separate.

2. Even if we decide the number of repos is manageable, it's not clear to me how we would achieve good code reuse when we have cloud-config fields embedded in types such as `KubeadmConfigSpec`. Example:

```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: KubeadmControlPlane
metadata:
  name: capi-azure-control-plane
  namespace: default
spec:
  kubeadmConfigSpec:
    ...
    diskSetup:
      filesystems:
      - device: /dev/disk/azure/scsi1/lun0
        extraOpts:
        - -E
        - lazy_itable_init=1,lazy_journal_init=1
        filesystem: ext4
        label: etcd_disk
      ...
      partitions:
      - device: /dev/disk/azure/scsi1/lun0
        layout: true
        overwrite: false
        tableType: gpt
```

In the snippet above, `diskSetup` is a field under `KubeadmConfigSpec`. This field is [specific to cloud-init](https://cloudinit.readthedocs.io/en/latest/reference/modules.html#disk-setup). That means `KubeadmConfigSpec` assumes cloud-init is the only provisioner. True, we could implement "translators" under the hood for Ignition and any other provisioner, but there is no guarantee for feature parity among all provisioners so IMO it's only a matter of time until we run into issues where some feature is supported by cloud-init but not by other provisioners.

I think we have a major design flaw in the current core API (mixing bootstrap and provisioning in the data model). True, it will be very disruptive to change this, but from my understanding you're concerned about building more complexity (e.g. Ignition 3) on top of a suboptimal foundation, and to me it seems we'd be doing exactly that if we keep the API as is and keep adding layers on top of it.

Also, I know that in the past some maintainers have expressed similar concerns: https://github.com/kubernetes-sigs/cluster-api/pull/4172#discussion_r600098540

To summarize, I think that if we avoid tackling the design problem head on and look for ways around it, we'd find ourselves in a similar situation later on, only then we'd be disrupting even more areas of the projects.

I'm open to changing my mind on this :slightly_smiling_face:

Reply:
@johananl that's a complex discussion
you clearly went miles down this path, and everyone else have to catch up (this is what the working group is for).
what I'm asking now is just to be open to other options and to not tight this working group to your work, because this will prevent other's contributions.

Also, please note that, I'm bringing this up mostly because, based on my experience, this could become an issue for the working group down the line when broader consensus will be required.

WRT to the initial assumption of the use cases we have in front of us, I get that theoretically we have to address a possible high number of combinations, but from the other side, as a matter of fact only the ignition team has raised this concern so far.

I'm not sure why, but I think that we should consider that either not all the combinations of providers/bootstrappers makes sense or simply there is no demand for many of those combinations (but getting a better understanding of the problem scope is again part of the working group goal IMO)

WRT to the possible solution, frankly speaking I think it is too early to get into the weed of different options without a better understanding of the context; also this PR is probably not the right place for those discussions.

But just to avoid misunderstanding, I'm well aware of the fact that the current KubeadmConfig API now is not in an ideal state for what we are discussing (as a matter of fact it wasn't designed with re-use in mind). However, what I'm advocating for is to explore options for making re-use possible while minimising the number of breaking changes/impact for users.

But finding a solution for this problem is a discussion that I will be happy to have if the charter of this working group opens up to such options.

Reply:
Hey @fabriziopandini. Thanks for your follow up.

Of course I'm open to discussing other solutions and would love to get contributions from others 🙂 I hope I didn't write anything which implies otherwise.

The specific solution in my WIP proposal definitely isn't the only one, and it also isn't a solution I had in mind in advance but rather what I could come up with after gathering a list of current issues and concerns from both users and maintainers around bootstrap and provisioning. I'm not attached to this proposal and am happy to drop it or replace it with a better proposal right away if we decide that's the way to go.

To be clear, the reason I became involved with this work initially was to improve the integration between Flatcar Container Linux and CAPI. So, with my Microsoft hat on I'd be very glad to avoid doing any major reworking of core CAPI pieces if it's not absolutely necessary. We'd be happy to simply keep Ignition supported, add support for Ignition v3 and work towards graduating Ignition support from experimental mode to a stable "GA".

However, I also have the OSS contributor hat, and with that hat on I always try to do what's right for the projects I'm involved in rather than solving ad-hoc issues which stand in my way while working on use cases relevant mainly to my company.

The reason I've gone into the weeds of #5294 and came up with this proposal is that previously we heard very clearly from some of the maintainers that before we can consider making Ignition a fully-supported, "GA" provisioner, we must tackle #5294. If that's no longer the case or if I somehow misunderstood the intention there -- great, as this would make our work around Ignition much shorter and easier 👍 

I'm also in favor of deferring the technical discussions to the work group.

I understand your point (as well as Stefan's) about keeping the API stable. That's very important indeed. The only reason I proposed changing core API objects is that as far as I understood the CAPI maintainers (or at least some of them) required that we cleanly separate bootstrap from provisioning, and I couldn't find a way to truly separate the two without touching the API because the two concerns are unfortunately entangled in the API.

Let's continue this discussion in the first iteration of the work group then 🙂

Reply:
great discussion here, i tend to agree with Fabrizio and Stefan about the need to not disrupt the API and user experience too much. with that said though, i think it would be great if the working group can explore multiple options to see if there is a way we can use the current APIs/contracts, or perhaps add to them instead of modifying, but as noted above this is technical discussion we should have in the WG.

for this proposal, i think we just need to be clear about the goals of the WG and our desire to find solutions that do not require re-architecting core pieces of the project.

Reply:
I've captured our discussion and included deep links to here in the WG agenda doc: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?usp=sharing

Will close this thread to ease review of this PR as aligned on in the [Nov 27 office hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).